packages/vite/src/node/plugins/resolve.ts

```diff
@@ -434,51 +446,54 @@ export function resolvePlugin(
             depsOptimizer,
             external,
             undefined,
+            this.environment.config,
           ))
         ) {
           return res
         }
 
-        // node built-ins.
-        // externalize if building for a node compatible environment, otherwise redirect to empty module
-        if (isBuiltin(id)) {
-          if (currentEnvironmentOptions.consumer === 'server') {
-            if (
-              options.noExternal === true &&
-              // if both noExternal and external are true, noExternal will take the higher priority and bundle it.
-              // only if the id is explicitly listed in external, we will externalize it and skip this error.
-              (options.external === true || !options.external.includes(id))
-            ) {
-              let message = `Cannot bundle Node.js built-in "${id}"`
-              if (importer) {
-                message += ` imported from "${path.relative(
-                  process.cwd(),
-                  importer,
-                )}"`
-              }
-              message += `. Consider disabling environments.${environmentName}.noExternal or remove the built-in dependency.`
-              this.error(message)
+        // built-ins
+        // externalize if building for a server environment, otherwise redirect to an empty module
+        if (
+          currentEnvironmentOptions.consumer === 'server' &&
+          this.environment.config.resolve.isBuiltin(id)
+        ) {
+          if (
+            options.noExternal === true &&
+            // if both noExternal and external are true, noExternal will take the higher priority and bundle it.
+            // only if the id is explicitly listed in external, we will externalize it and skip this error.
+            (options.external === true || !options.external.includes(id))
```

Comment:
`resolve.isBuiltin` and `resolve.external` and `resolve.noExternal` interacts here. I tried to write down the use case for each settings:

| `external` | `noExternal` | `isBuiltin` | use case |
|-----------|--------------|------------|-----------|
| []             | []                 | isNodeBuiltin | for applications that runs on Node |
| []             | [...]               | isNodeBuiltin | for applications that runs on Node and wants to bundle some deps |
| [...]           | [...]               | isNodeBuiltin | for applications that runs on Node and wants to bundle some deps |
| []             | true             | isNodeBuiltin | for applications that runs on Node and wants to bundle all deps |
| [...] / true | true             | isNodeBuiltin | for applications that runs on non-Node compat runtime and wants to bundle all deps |
| []             | []                 | customFunction | ? (I guess this won't work. Dependencies will be externalized but if the runtime is not Node compatible then many deps won't work) |
| []             | [...]               | customFunction | ? (same with above) |
| [...]           | [...]               | customFunction | ? (same with above) |
| []             | true             | customFunction | for applications that runs on non-Node compat runtime and wants to bundle all deps (What is the advantage of it compared to `external: [...], noExternal: true, isBuiltin: isNodeBuiltin`?) |
| [...] / true | true             | customFunction | for applications that runs on non-Node compat runtime and wants to bundle some deps (What is the advantage of it compared to `external: [...], noExternal: true, isBuiltin: isNodeBuiltin`?) |

Do we have use cases for the ones with `?`? I wonder if the option should be part of `resolve.external` (something like `external: { isBuiltin: [], dependencies: [] }`). or maybe making `resolve.external` work better could work?

Reply:
What do you mean by `I guess this won't work. Dependencies will be externalized but if the runtime is not Node compatible then many deps won't work`?

If I have a runtime that does not support node builtin modules (or only a subset of them) like for example workerd without the nodejs compat flag turned on, it seems ok that "many deps won't work" since they would also not work in the final deployed environment, no?

Reply:
> or maybe making resolve.external work better could work?

Yes... maybe 🤔 but improving `resolve.external` won't help the fact that the current `isBuiltin` function is hardcoded for (more or less) node.

What I mean is, let's say that user code is importing `node:path` in a workerd worker without setting the nodejs compat flag, currently there is no way to tell vite that `node:path` is not a builtin module of the workerd environment, is there?

Reply:
> What do you mean by `I guess this won't work. Dependencies will be externalized but if the runtime is not Node compatible then many deps won't work`?
> 
> If I have a runtime that does not support node builtin modules (or only a subset of them) like for example workerd without the nodejs compat flag turned on, it seems ok that "many deps won't work" since they would also not work in the final deployed environment, no?

I'm not sure why I wrote that. Probably I mixed up something. Please ignore it 😅

> What I mean is, let's say that user code is importing `node:path` in a workerd worker without setting the nodejs compat flag, currently there is no way to tell vite that `node:path` is not a builtin module of the workerd environment, is there?

If `noExternal` is set to `true` and `external` does not include node builtins, it should work like that (https://github.com/vitejs/vite/pull/15656).

Reply:
> If noExternal is set to true and external does not include node builtins

Yeah I double checked that with our plugin, it does work as you said 👍